### ISSUE TITLE
[macOS] Inconsistent focus in branch-creation and branch-renaming modals

### ISSUE THREAD

**Issue by yakovmanshin** [desktop/desktop#17661]
### The problem

When you create a new branch, the modal window automatically focuses on the Name field.

![Branch Creation](https://github.com/desktop/desktop/assets/4648742/4fc38982-9f93-46ed-811a-bf8b9e564538)

However, when you rename a branch by using the <kbd>shift</kbd> + <kbd>command</kbd> + <kbd>R</kbd> shortcut, focus is set on the Close button. This requires a mouse click or <kbd>Tab</kbd> press before you can type the new name.

![Branch Renaming (Keyboard Shortcut)](https://github.com/desktop/desktop/assets/4648742/ae9abf14-bd2a-42cc-9a76-f7744984ae9b)

And when you rename a branch by selecting Rename… from the context menu, focus doesn’t appear to be set at all. Technically, it’s still on the Close button (you can see it momentarily highlight if you press <kbd>return</kbd>).

![Branch Renaming (Mouse)](https://github.com/desktop/desktop/assets/4648742/e1959ef6-194e-48b1-acc8-8c56f7d45cb7)

While it’s not a major issue, this inconsistency feels confusing and slows down the work (especially if you’re used to navigating the app with keyboard shortcuts).

### Release version

3.3.5

### Operating system

macOS 14

### Additional context

This issue first appeared a few weeks ago. Previously, focus was automatically set on the text field everywhere.

**Comment by tidy-dev** [comment]
This was done for an accessibility concern because sometimes the rename branch has a warning about being associated to a remote branch before the input. Thus, if we focus the input first, screen reader users would not be made aware of this with logical forward progression through the dialog.  We have updated our focus logic such that if there is no skippable content before the first input, it will receive focus. 

As evidenced in this issue, 🙂 sometimes the rename dialog does not have a warning. We could update it such that it only moves to the "x" if a warning is present.

Other notes: 
- Moving the warning beneath the input was previously discussed and I believe the thought there was that a warning that is not specific to an input should not be directly beneath an input since it does not directly relate. This warning is about "renaming" in general, it is a dialog warning, not input warning. Thus, here the warning is given first because it is about what the action of renaming will do and not specifically with what the chose input in the rename will do.  Not sure if putting the warnings in the footer was considered... we do that in the merge/rebase dialog and the preview pull request?
- Another option discussed was to make the warning a confirmation, but I think it was agreed that the warning was not a strong enough problem to require a confirmation.

**Comment by yakovmanshin** [comment]
Thanks for your clarification @tidy-dev!

I just tested branch renaming with VoiceOver and noticed the initial announcement varies based on whether the modal was opened with the shortcut (“dialog, with 3 items, Close, button”) or mouse (“Closing menu combo box, dialog, with 3 items, Close, button”).

I’m not very familiar with accessibility features of Electron, but VoiceOver supports extra accessibility metadata that’s not visible in the UI. I imagine hearing something like “Rename Branch—warnings ahead, dialog, with 3 items…” would solve the problem with warning awareness—plus it’s just helpful to hear the window’s title rather than a generic “Closing menu combo box.” Do you think a change like this can be implemented?

**Comment by tidy-dev** [comment]
❤  Thanks for testing!

> “dialog, with 3 items, Close, button” .... it’s just helpful to hear the window’s title

Normally the dialog title is announced there. There was a recent regression with VoiceOver that made it so our dialog titles are no longer announced. But, a workaround fix for it is in our current beta. 

> “Rename Branch—warnings ahead

Interesting idea. However, we are trying to follow established accessibility patterns and the specs set out in guidelines such as WCAG. That way GitHub Desktop accessibility behavior is predictable to users of any manner of accessibility software. 

The suggestions I made I know follow those guidelines. (Moving warning to footer, or only focusing `x` when warning is present)

Other thoughts that would need conference with our accessibility team:
- Could this dialog when it has warnings be elevated to an `alertdialog` ? Alert dialogs announce the title along with a description (which could be the warning). Usually they are used for errors and confirmation dialogs -> content that we definitely want the user to acknowledge before proceeding. Then, since the warning is announced, maybe the focus could be placed on input, even tho that is still skipping content physically.  Thinking this could still be considered an issue for users of zoom software that may be zoomed into the point where they would not see the warning above.
- Could this dialog have a warning banner with a role of alert. We have error banners. I believe tho the proper pattern with that would still be that the banner would receive focus first... but, maybe at least then, it would be understandable why the input is not focused and the input would only not be focused when such banner exists.
